src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs

```diff
@@ -794,7 +807,20 @@ public virtual void AddGacReference(string name)
                     assemblyFullPath = Path.Combine(projectFullPath, referencePath);
 
                     // Add a reference to the project
-                    dynamic reference = References.Add(assemblyFullPath);
+                    dynamic reference;
+                    try
+                    {
+                        // First try the References3.AddFiles API, as that will incur fewer
+                        // design-time builds.
+                        References3.AddFiles(new[] { assemblyFullPath }, out var referencesArray);
```

Comment:
Shouldn't we now remove setting CopyLocal explicitly? Look at line no 840

Reply:
If we fall back to `References.Add` we still need to set CopyLocal explicitly.

Reply:
hmm i see, will there be any perf impact of setting this CopyLocal for `Reference3`?

Reply:
No, since `TrySetCopyLocal` first checks the `CopyLocal` value, and only sets it if it is not `true`. When using `References3.AddFiles` it will always be `true`.